### ISSUE TITLE
Scoped styles library WIP problems

### ISSUE THREAD

**Issue by theSherwood** [luwes/sinuous#92]
Hi @luwes. I don't know if you've got time for this, but I'm throwing together a library for sinuous that is kind of like [styled-jsx](https://github.com/zeit/styled-jsx). Here's a sandbox where you can see what I've got going: [sandbox](https://codesandbox.io/s/sinuous-scoped-style-wq5fo). I'm running into a couple problems that seem to be each other's solutions but I can't sort out how. If you have the time, I'd be grateful for your help.

Library strategy:
Intercept calls to `html` and `api.h` in order to add classes to every node and scope styles to those classes. I also check to see if there is already the style element on the page to avoid adding it for every time the component is rendered. This requires rendering it outside where it would have otherwise rendered (at the moment I'm simply appending it to the body, though this will likely change).

Problem 1:
If all the components are removed from the page and then added back on (such as can happen when updating a list), the dynamic style elements (style elements that render an observable) stop working. They somehow lose connection to the observables.

Problem 2:
If all the scoped-style components are removed from the page, I would like to remove the style elements from the page.

The two problems hint at being the solution to each other's problems. If I can detect when the observables lose connection to the style elements (Problem 1), I can remove the style elements from the page (solving Problem 2), which will mean that when components are added back to the page, I can also add the style elements again (solving Problem 1).

If you have any insight into this, I'd really appreciate it. I think this could be a handy library for expanding sinuous' capabilities when being written and run directly in the browser.

**Comment by luwes** [comment]
Nice! This might be an easy fix, have you tried using `sinuous/map` for the list?

**Comment by theSherwood** [comment]
I could do in this case. But there may be times when all the instances of a certain component are genuinely removed from the dom regardless of how that happens. Using the list the way I am is just a convenient way to experiment with that. Line 21 of app.js prevents Problem 1 by preventing all instances of the `test` component from being removed from the list. But if you delete that line, then click the `push` button, the "Make Gray"/"Make Black" and "Make Heavy"/"Make Light" cease to work. The components that get added to the page maintain their connection to the observable, but the style component loses that connection. I'm guessing it has something to do with cleanup. But do you have any ideas what might be causing it?

**Comment by theSherwood** [comment]
You know what, this example isn't great. I'll work on simplifying the example as much as I can to make it easy to see the problems. This sandbox is too messy.

**Comment by luwes** [comment]
ah I see, this has to do with the parent computations cleaning up their children.
if you want to make a child work standalone, it needs to be wrapped around a root.
then the child can be removed and later re-added and the connection will still be there. the downside of this is that you'll have to manage cleaning up the child computations manually. 

S.js has great docs around this, which mostly apply to sinuous/observable as well
https://github.com/adamhaile/S#srootdispose--code

**Comment by theSherwood** [comment]
Thanks! That solved one of the problems! Turns out the style component was getting disconnected from the observable on the cleanup of the first component instance. Using `root` solved that. 

So that leaves Problem 2. Is there any way that you can think of to detect when all instances of a component have been removed from the page? 

Here is a [simplified example sandbox](https://codesandbox.io/s/sinuous-scoped-style-simplified-example-tj3q8). You can toggle the components on and off, and after the initial render, the number of style components will not have changed. Which is what is desired. However, it would be great to be able to detect when all the components have been removed from the page so that the style elements can also be removed. I'm scratching my head as to how to do that but can't come up with much. Any ideas?

**Comment by ryansolid** [comment]
Couldn't manage a counter with an initialization and cleanup hook? You control the call point so you just need a registry object that exists globally and every time you create you either set a new key at 1 or you add 1 to an existing. Everytime a cleanup call in the same context happens decrement or delete if it's the last one.

**Comment by theSherwood** [comment]
@ryansolid Are you talking about the cleanup of observables/computeds? I have been trying to think through cleaning up through some of the dom inserts and removals, but that doesn't really seem to work. I hadn't considered registering cleanups on every call that returns a dom element. Am I understanding that right? Wouldn't that be pretty expensive?

**Comment by ryansolid** [comment]
Maybe I'm not understanding the problem correctly but your styled context lives within your single template literal call. So in your "2 component" example, there are exactly 2 calls happening (unless you press toggle). One for each component. And they only run when you run that component for the first time(or toggle on). I haven't looked at the implementation or what Sinuous provides so I could be way off base, but I'm thinking somewhere you know you are looking at that style tag and you know not to add it twice so wherever you add it to the head is the place to do it: 

```
const styleRegistry = { }

// assuming the suggested syntax from the issue
function shtml(styleHash) {
  return (template, ...args) {
    /* somewhere we know we are creating style tag grab css expression */
    if (!styleRegistry[styleHash]) {
      styleRegistry[styleHash] = 1;
      appendCSSToHead(styleHash, css)
    } else styleRegistry[styleHash]++;
    cleanup(() => {
      if (--styleRegistry[styleHash] < 1) {
        delete styleRegistry[styleHash];
        removeCSSFromHead(styleHash);
      }
    })
  }
}

```

**Comment by theSherwood** [comment]
@ryansolid Yep! That solved it. As of version 0.3.1, sinuous-style cleans up unused styles. 

Thanks for pointing that out to me. I'm still trying to wrap my head around the way the reactive system integrates with the rendering. Doing something like the above had never occurred to me!

**Comment by ryansolid** [comment]
The problem is I think the nested render model is still not clear to everyone. I need to write a real detailed article about this. I always stop at the reactive system but it's important. It is pretty common for people to look for component lifecycles and DOM events etc, but the reactive tree handles this all for you.. No matter where you are in the app (assuming it's in a root) even if it is not tracked you are inside a computed. So if there is no conditional between you and that root you are always going to be rendered. You will live the life of the app. In order for an arbitrary nested component to be added or removed there needs to be a wrapping reactive context--something to re-evaluate to render or remove your component. Ideally, this context doesn't keep triggering because you'd be wasting a lot of work recreating everything (since we don't virtually diff). So assuming you are doing things correctly, that means your component's parent context will only re-run when it is being removed. If it is re-run more it is essentially the same as removing and creating anyway since you will be making all new DOM nodes. So for this reason a `cleanup` method is basically as good as an unmount method. The only only exception is if you hoist the DOM nodes out of the reactive context that they are rendered in (like caching an offscreen tab). Then they will belong to the parent context and won't be released until you remove the whole tab bar.

This isn't obvious at first though. We don't want to call the method `unmount` because it can also be put in computations as a way to cleanup side effects. But it's all really the same thing since our DOM nodes are side effects of the reactive system. So it's incredibly simple and incredibly powerful. Yet completely not obvious because of how we've been trained. In my opinion, it is actually quite clean and elegant.

**Comment by theSherwood** [comment]
That's really clarifying. For some reason it is hard for me to remember that wherever you are in the app, you are within some reactive context. There's something that's almost too magical about it, that it's just lurking in the background all the time.  

> In my opinion, it is actually quite clean and elegant.

The primitives and patterns are very simple and general. I think that is part of why it is hard to remember what you can do with it. Other frontend libraries have pretty clear boundaries around what their primitives and patterns can and cannot do. But this reactive pattern creates a whole new landscape that a program exists within. 

Long story short, these patterns are fascinating, but I feel like I need a diagram or something to help me get a better understanding of the big picture (and also specifically how it interacts with the render engine).
